README
README file for Tree::Simple.

See also: CHANGES and Changelog.ini.

Warning: WinZip 8.1 and 9.0 both contain an 'accidental' bug which stops
them recognizing POSIX-style directory structures in valid tar files.
You are better off using a reliable tool such as InfoZip:
ftp://ftp.info-zip.org/pub/infozip/

1 Installing from a Unix-like distro
------------------------------------
shell>gunzip Tree-Simple-1.23.tgz
shell>tar mxvf Tree-Simple-1.23.tar

On Unix-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>./Build
shell>./Build test
shell>./Build install

On MS Windows-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>perl Build
shell>perl Build test
shell>perl Build install

Alternately, without Module::Build, you do this:

Note: 'make' on MS Windows-like systems may be called 'nmake' or 'dmake'.

shell>perl Makefile.PL
shell>make
shell>make test
shell>su              (for Unix-like systems)
shell>make install
shell>exit            (for Unix-like systems)

On all systems:

Run Simple.pm through your favourite pod2html translator.

2 Installing from an ActiveState distro
---------------------------------------
shell>unzip Tree-Simple-1.23.zip
shell>ppm install --location=. Tree-Simple
shell>del Tree-Simple-1.23.ppd
shell>del PPM-Tree-Simple-1.23.tar.gz


Changes
Revision history for Perl extension Tree::Simple.

1.26  2015-11-28T09:30:00
	- No code changes.
	- Accept patches from Manwar via github. See https://github.com/stevan/tree-simple/pull/10.
	- I (Ron) followed github's advice on handling merge conflicts from the command line, but could
		not see the patches after the merge.
	- Oops. Can't get Test::Stream to work in 10_Tree_Simple_test.t. Logged issue, but it's
		probably a problem with the way I tried to use Test::Stream. Revert to Test::More.
	- Fix reference to github repo in Makefile.PL. Part was missing.
	- Add xt/author/changes.t to MANIFEST.

1.25  2014-01-13T08:51:00
	- Fix logic in sub DESTROY for when children are/are not present. Reported by astortz.
		See https://github.com/stevan/tree-simple/issues/8.

1.24  2014-09-06T09:27:00
	- Abandon Test::Version and hence delete t/version.t. Putting version.t in xt/author/ is not a
		solution, because Test::Version has problems with Test::EOL and Test::Builder.
		Thanx to Kent Fredric for the github issue which started me investigating this issue.

1.23  2013-11-09T10:43:00
	- No code changes.
	- Accept spelling corrections from the Debian Perl Group, with thanx. See RT#90171.
		Note: The change to lib/Tree/Simple/Visitor.pm had already been made.
	- Add t/version.t, which uses Test::Version, to cross-check version #s.

1.22  2013-09-30T08:35:00
	- Metadata fixes in Build.PL and correct bugtracker queue in Makefile.PL. Thanx to dsteinbrunner
		(via github).

1.21  2013-09-26T13:08:00
	- Fix syntax error in Makefile.PL.

1.20  2013-09-26T11:16:00
	- Merge source from https://github.com/stevan/tree-simple.
	- Close github issues.
	- Add bugtracker and license to Build.PL and Makefile.PL.

1.19  2013-09-23T08:26:00
	- Maintenance now by Ron Savage.
	- Rectify datestamp format in this file.
	- Add Changelog.ini.
	- Make Simple.pm and Visitor.pm have the same version #.
	- Reformat Build.PL and Makefile.PL.
	- Move t/pod* into xt/author/.
	- Minor doc patches.
	- Add t/17_Tree_Simple_width_test.t. This code was attached to RT#40407, to demonstrate a bug in
		keeping a running total of the width (leaf count) of a tree during node additions to leaves.
		Thanx to David Cryer for the test code and patch.
	- Add t/21_Tree_Simple_Visitor_test.t. This code was attached to RT#30032, to demonstrate a bug
		in including the root in a traversal after calling $visitor -> includeTruck(1).
	- RT#30032: Adopt patch from Moses Amaro. With thanx.
	- RT#38607: Reject. Suppressing deep recursion warnings should not normally be done. Sub-class!
	- RT#40407: Adopt patch from David Cryer. With thanx.
	- RT#84797: Reject. Changing the return value of setUID could break any amount of code.

1.18  2007-11-11T12:00:00
	- fixing version string to not choke on 5.10 (RT #29746).

1.17  2006-10-23T12:00:00
	- Make loading of Scalar::Util::weaken, completely optional
	- Added a $post_func optional arg to &traverse. Thanks to Mark Lawrence for the patch, docs and
		tests :).

1.16  2006-02-06T12:00:00
	- Converted to use Module::Build (Rob Kinyon)
	- Refactored &addChild and &addChildren to be implemented in terms of &insertChild and
		&insertChildren (Rob Kinyon)
	- Other misc. refactorings (Rob Kinyon)
	- Updated Scalar::Util version dependency (Stevan Little)
	- Updated copyrights for the new year (Stevan Little)

1.15  2005-05-26T12:00:00
	- Added optional use of weakened parent references and improved the old circular reference
		DESTROY model to make more sense. See the documentation for more info.
	- Fixed bug in the fixDepth() function.

1.14  2004-11-18T12:00:00
	- Now using Scalar::Util::blessed() instead of the convoluted UNIVERSAL::isa() stuff.
	- Added Scalar::Util as a dependency.

1.13  2004-11-15T12:00:00
	- Added width functionality (with getWidth). Thanks to Mark Thomas for his insight/suggestions.
	- Added tests for this.
	- Added documentation for this.
	- Improved the height functionality, thanks again to Mark Thomas for his insight/suggestions.
	- Deprecated the 'height' method in favor of the more consistent 'getHeight' method.
	- Added tests for this.
	- Added documentation for this.
	- Added some info in the docs to explain the depth value for ROOT trees.
	- Cleaned up and improved the following test files:
		11_Tree_Simple_fixDepth_test.t
		13_Tree_Simple_clone_test.t

1.12  2004-10-07T12:00:00
	- Fixed the clone method, it was not correctly cloning parental relationships.
	- Added tests and docs for this.
	- Improved clone and cloneShallow with the addition of the _cloneNode utility function, we now
		deep clone the nodes.
	- Added test and docs for this.

1.11  2004-10-04T12:00:00
	- Some documentation changes, no code changes.

1.10  2004-08-31T12:00:00
	- Streamlined the DESTROY method to avoid method calls as this can sometimes cause issues
		during global destruction with subclasses.

1.09  2004-08-31T12:00:00
	- Fixed DESTROY to avoid memory leaks (RT-BUG: #7512).
	- Added documentation to explain when to call the DESTROY method to properly clean up any
		circular references.
	- Added test (14_Tree_Simple_leak_test.t) to verify this fix. Needs Test::Memory::Cycle to run.

1.08  2004-08-25T12:00:00
	- Added the 'height' and 'size' methods.
	- Added tests for these.
	- Added documentation for these.

1.07  2004-07-28T12:00:00
	- Added the getUID and setUID methods to Tree::Simple. Thanks to Brett Nuske for that
		suggestion.
	- Added documentation for these methods.
	- Added tests for those methods.
	- Added t/pod.t and t/pod_coverage.t to the test suite.

1.06  2004-07-06T12:00:00
	- Changed what the Tree::Simple accept method accepts.
	- Added tests for this.
	- Completely Revised the Tree::Simple::Visitor object so that it is a better base class.
		This coincides with the release of the Tree::Simple::VisitorFactory collection of Visitor
		objects.
	- Added tests for this.

1.05  2004-06-06T12:00:00

Makefile.PL
use 5.006;
use strict;
use warnings;

use ExtUtils::MakeMaker;

# ----------------

my(%params) =
(
	($] ge '5.005') ?
	(
		AUTHOR   => 'Stevan Little <stevan@cpan.org>',
		ABSTRACT => 'A simple tree object',
	) : (),
	clean =>
	{
		FILES => 'blib/* Makefile Tree-Simple-*'
	},
	dist =>
	{
		COMPRESS => 'gzip',
		SUFFIX   => 'gz'
	},
	DISTNAME  => 'Tree-Simple',
	NAME      => 'Tree::Simple',
	PL_FILES  => {},
	PREREQ_PM =>
	{
		'constant' => 0,
		'Scalar::Util' => '1.18',
		'strict' => 0,
		'Test::Exception' => '0.15',
		'Test::More' => 1.001014,
		'warnings' => 0,
	},
	VERSION_FROM => 'lib/Tree/Simple.pm',
);

if ( ($ExtUtils::MakeMaker::VERSION =~ /^\d\.\d\d$/) && ($ExtUtils::MakeMaker::VERSION > 6.30) )
{
	$params{LICENSE} = 'artistic_2';
}

if ($ExtUtils::MakeMaker::VERSION ge '6.46')
{
	$params{META_MERGE} =
	{
		resources =>
		{
			bugtracker => 'https://rt.cpan.org/Public/Dist/Display.html?Name=Tree-Simple',
			license    => 'http://www.perlfoundation.org/artistic_license_2_0',
			repository =>
			{
				type => 'git',
				url  => 'https://github.com/ronsavage/Tree-Simple.git',
				web  => 'https://github.com/ronsavage/Tree-Simple',
			},
		},
	};
}

WriteMakefile(%params);


lib/Tree/Simple/Visitor.pm
package Tree::Simple::Visitor;

use strict;
use warnings;

our $VERSION = '1.26';

use Scalar::Util qw(blessed);

## class constants

use constant RECURSIVE     => 0x01;
use constant CHILDREN_ONLY => 0x10;

### constructor

sub new {
	my ($_class, $func, $depth) = @_;
	if (defined($depth)){
		($depth =~ /\d+/ && ($depth == RECURSIVE || $depth == CHILDREN_ONLY))
                    || die "Insufficient Arguments : Depth arguement must be either ".
                    "RECURSIVE or CHILDREN_ONLY";
	}
	my $class = ref($_class) || $_class;
    # if we have not supplied a $func
    # it is automatically RECURSIVE
    $depth = RECURSIVE unless defined $func;
	my $visitor = {
		depth => $depth || 0
		};
	bless($visitor, $class);
	$visitor->_init();
    if (defined $func) {
        $visitor->setNodeFilter($func);
        $visitor->includeTrunk(1);
    }
	return $visitor;
}

### methods

sub _init {
	my ($self) = @_;
    $self->{_include_trunk} = 0;
    $self->{_filter_function} = undef;
    $self->{_results} = [];
}

sub includeTrunk {
    my ($self, $boolean) = @_;
    $self->{_include_trunk} = ($boolean ? 1 : 0) if defined $boolean;
    return $self->{_include_trunk};
}

# node filter methods

sub getNodeFilter {
    my ($self) = @_;
	return $self->{_filter_function};
}

sub clearNodeFilter {
    my ($self) = @_;
	$self->{_filter_function} = undef;
}

sub setNodeFilter {
    my ($self, $filter_function) = @_;
	(defined($filter_function) && ref($filter_function) eq "CODE")
            || die "Insufficient Arguments : filter function argument must be a subroutine reference";
	$self->{_filter_function} = $filter_function;
}

# results methods

sub setResults {
    my ($self, @results) = @_;
    $self->{results} = \@results;
}

sub getResults {
    my ($self) = @_;
    return wantarray ?
             @{$self->{results}}
             :
             $self->{results};
}

# visit routine
sub visit {
	my ($self, $tree) = @_;
	(blessed($tree) && $tree->isa("Tree::Simple"))
            || die "Insufficient Arguments : You must supply a valid Tree::Simple object";
    # get all things set up
	my @results;
	my $func;
    if ($self->{_filter_function}) {
        $func = sub { push @results => $self->{_filter_function}->(@_) };
    }
    else {
        $func = sub { push @results => $_[0]->getNodeValue() };
    }
	# always apply the function
	# to the tree's node

    $func->($tree) if (defined($self->{_include_trunk}) && $self->{_include_trunk});

	# then recursively to all its children
	# if the object is configured that way
	$tree->traverse($func) if ($self->{depth} == RECURSIVE);
	# or just visit its immediate children
	# if the object is configured that way
	if ($self->{depth} == CHILDREN_ONLY) {
		$func->($_) foreach $tree->getAllChildren();
	}
    # now store the results we got
    $self->setResults(@results);
}


1;

__END__

=head1 NAME

Tree::Simple::Visitor - Visitor object for Tree::Simple objects

=head1 SYNOPSIS

  use Tree::Simple;
  use Tree::Simple::Visitor;

  # create a visitor instance
  my $visitor = Tree::Simple::Visitor->new();

  # create a tree to visit
  my $tree = Tree::Simple->new(Tree::Simple->ROOT)
                         ->addChildren(
                             Tree::Simple->new("1.0"),
                             Tree::Simple->new("2.0")
                                         ->addChild(
                                             Tree::Simple->new("2.1.0")
                                             ),
                             Tree::Simple->new("3.0")
                             );

  # by default this will collect all the
  # node values in depth-first order into
  # our results
  $tree->accept($visitor);

  # get our results and print them
  print join ", ", $visitor->getResults();  # prints "1.0, 2.0, 2.1.0, 3.0"

  # for more complex node objects, you can specify
  # a node filter which will be used to extract the
  # information desired from each node
  $visitor->setNodeFilter(sub {
                my ($t) = @_;
                return $t->getNodeValue()->description();
                });

  # NOTE: this object has changed, but it still remains
  # backwards compatible to the older version, see the
  # DESCRIPTION section below for more details

=head1 DESCRIPTION

This object has been revised into what I think is more intelligent approach to Visitor objects.
This is now a more suitable base class for building your own Visitors. It is also the base class
for the visitors found in the B<Tree::Simple::VisitorFactory> distribution, which includes a
number of useful pre-built Visitors.

While I have changed a number of things about this module, I have kept it backwards compatible
to the old way of using it. So the original example code still works:

  my @accumulator;
  my $visitor = Tree::Simple::Visitor->new(sub {
                        my ($tree) = @_;
                        push @accumulator, $tree->getNodeValue();
                        },
                        Tree::Simple::Visitor->RECURSIVE);

  $tree->accept($visitor);

  print join ", ", @accumulator;  # prints "1.0, 2.0, 2.1.0, 3.0"

But is better expressed as this:

  my $visitor = Tree::Simple::Visitor->new();
  $tree->accept($visitor);
  print join ", ", $visitor->getResults();  # prints "1.0, 2.0, 2.1.0, 3.0"

This object is still pretty much a wrapper around the Tree::Simple C<traverse> method,
and can be thought of as a depth-first traversal Visitor object.

=head1 METHODS

=over 4

=item B<new ($func, $depth)>

The new style interface means that all arguments to the constructor are now optional.
As a means of defining the usage of the old and new, when no arguments are sent to the
constructor, it is assumed that the new style interface is being used. In the new style,
the C<$depth> is always assumed to be equivalent to C<RECURSIVE> and the C<$func> argument
can be set with C<setNodeFilter> instead. This is the recommended way of doing things now.
If you have been using the old way, it is still there, and I will maintain backwards
compatibility for a few more version before removing it entirely. If you are using this
module (and I do not even know if anyone actually is) you have been warned. Please contact
me if this will be a problem.

The old style constructor documentation is retained her for reference:

The first argument to the constructor is a code reference to a function which expects a
B<Tree::Simple> object as its only argument. The second argument is optional, it can be
used to set the depth to which the function is applied. If no depth is set, the function
is applied to the current B<Tree::Simple> instance. If C<$depth> is set to C<CHILDREN_ONLY>,
then the function will be applied to the current B<Tree::Simple> instance and all its
immediate children. If C<$depth> is set to C<RECURSIVE>, then the function will be applied
to the current B<Tree::Simple> instance and all its immediate children, and all of their
children recursively on down the tree. If no C<$depth> is passed to the constructor,
then the function will only be applied to the current B<Tree::Simple> object and none
 of its children.

=item B<includeTrunk ($boolean)>

Based upon the value of C<$boolean>, this will tell the visitor to collect the trunk of the
tree as well. It is defaulted to false (C<0>) in the new style interface, but is defaulted
to true (C<1>) in the old style interface.

=item B<getNodeFilter>

This method returns the CODE reference set with C<setNodeFilter> argument.

=item B<clearNodeFilter>

This method clears node filter field.

=item B<setNodeFilter ($filter_function)>

This method accepts a CODE reference as its C<$filter_function> argument. This code reference
is used to filter the tree nodes as they are collected. This can be used to customize output,
or to gather specific information from a more complex tree node. The filter function should
accept a single argument, which is the current Tree::Simple object.

=item B<getResults>

This method returns the accumulated results of the application of the node filter to the tree.

=item B<setResults>

This method should not really be used outside of this class, as it just would not make any sense
to. It is included in this class and in this documentation to facilitate subclassing of this
class for your own needs. If you desire to clear the results, then you can simply call
C<setResults> with no argument.

=item B<visit ($tree)>

The C<visit> method accepts a B<Tree::Simple> and applies the function set in C<new> or
C<setNodeFilter> appropriately. The results of this application can be retrieved with C<getResults>

=back

=head1 CONSTANTS

These constants are part of the old-style interface, and therefore will eventually be deprecated.

=over 4

=item B<RECURSIVE>

If passed this constant in the constructor, the function will be applied recursively down the
hierarchy of B<Tree::Simple> objects.

=item B<CHILDREN_ONLY>

If passed this constant in the constructor, the function will be applied to the immediate children
of the B<Tree::Simple> object.

=back

=head1 BUGS

None that I am aware of. The code is pretty thoroughly tested (see B<CODE COVERAGE> section in
B<Tree::Simple>) and is based on an (non-publicly released) module which I had used in production
systems for about 2 years without incident. Of course, if you find a bug, let me know, and I will
be sure to fix it.

=head1 SEE ALSO

I have written a set of pre-built Visitor objects, available on CPAN as
B<Tree::Simple::VisitorFactory>.

=head1 AUTHOR

stevan little, E<lt>stevan@iinteractive.comE<gt>

=head1 REPOSITORY

L<https://github.com/ronsavage/Tree-Simple>.

=head1 COPYRIGHT AND LICENSE

Copyright 2004-2006 by Infinity Interactive, Inc.

L<http://www.iinteractive.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


Build.PL
use 5.006;
use strict;
use warnings;

use Module::Build;

# ----------------

my $class = Module::Build->subclass(code => <<'EOF');

sub ACTION_authortest
{
	my($self) = @_;

	$self->depends_on('build');
	$self->depends_on('manifest');
	$self->depends_on('distmeta');

	$self->test_files( qw< t xt/author > );
	$self->recursive_test_files(1);

	$self->depends_on('test');

	return;
}
EOF

$class -> new
(
	dist_abstract  => 'A simple tree object',
	dist_author    => 'Stevan Little <stevan@cpan.org>',
	license        => 'perl',
	module_name    => 'Tree::Simple',
	build_requires =>
	{
		'Test::Exception' => '0.15',
		'Test::More' => 1.001014,
	},
	configure_requires =>
	{
		'Module::Build' => 0.4211,
	},
	requires =>
	{
		'constant' => 0,
		'Scalar::Util' => '1.18',
		'strict' => 0,
		'warnings' => 0,
	},
	meta_merge => {
		resources => {
		'repository' => 'https://github.com/stevan/tree-simple',
		'bugtracker' => 'https://github.com/stevan/tree-simple/issues',
		'license' => 'https://www.gnu.org/licenses/license-list.html#PerlLicense',
		},
	},
) -> create_build_script;
